### PR TITLE
feat: add Vec support as an Input

### LINKED ISSUE #434: Add `Vec<T>` support
### Description
Currently, the SDK only supports static-sized arrays in the format [T; n] where T is the type and n is the length.
Soon, Sway will release Vec, a growable/mutable/dynamic vector. We need to support this on the SDK side.

Details & Blockers: https://github.com/FuelLabs/fuels-rs/issues/353

### Initial checklist
- [ ] Enable to call a contract fn passing a Vec<T>
- [ ] Call a contract fn and receive a Vec<T> as return
- [ ] Generate the correct `Type` using `Typegen` for the `Vec<T>`

### PR DESCRIPTION
**Pull request by camsjams**
Closes #434 

Here is a basic example of my thought process, also see #434 for greater details
Blank diagram

### CODE REVIEW

File: packages/abi-coder/src/abi-coder.ts [hunk 1]
```diff
@@ -16,18 +16,60 @@ import StringCoder from './coders/string';
...
+type ByteInfo = { vecByteLength: number } | { byteLength: number };
+
+const getVectorAdjustments = (coders: Coder<unknown, unknown>[], values: InputValue[]) => {
+  const vectorData: Uint8Array[] = [];
+  const byteMap: ByteInfo[] = coders.map((encoder, i) => {
+    if (!(encoder instanceof VecCoder)) {
+      return { byteLength: encoder.encodedLength };
+    }
+
+    const data = encoder.getEncodedVectorData(values[i] as any);
```
Comment (luizstacio) [inline packages/abi-coder/src/abi-coder.ts:42]: Can you add a comment to ignore any here?

File: packages/abi-coder/src/abi-coder.ts [hunk 2]
```diff
@@ -16,18 +16,60 @@ import StringCoder from './coders/string';
 import StructCoder from './coders/struct';
 import TupleCoder from './coders/tuple';
 import U64Coder from './coders/u64';
+import VecCoder from './coders/vec';
```
Comment (luizstacio) [inline packages/abi-coder/src/abi-coder.ts:19]: We should also include `export *  from ./coders/vec` on the abi-coder index.ts file

File: packages/abi-coder/src/abi-coder.ts [hunk 3]
```diff
@@ -16,18 +16,60 @@ import StringCoder from './coders/string';
...
   OPTION_CODER_TYPE,
+  VEC_CODER_TYPE,
 } from './constants';
 import type { JsonAbiFragmentType } from './json-abi';
 import { filterEmptyParams, hasOptionTypes } from './utilities';
 
 const logger = new Logger(process.env.BUILD_VERSION || '~');
+type ByteInfo = { vecByteLength: number } | { byteLength: number };
+
+const getVectorAdjustments = (coders: Coder<unknown, unknown>[], values: InputValue[]) => {
```
Comment (luizstacio) [inline packages/abi-coder/src/abi-coder.ts:35]: Should this function be moved to `./coders/vec` or `./utilities`?

File: packages/abi-coder/src/coders/vec.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  getEncodedVectorData(value: InputValueOf<TCoder>): Uint8Array {
+    if (!Array.isArray(value)) {
+      this.throwError('expected array value', value);
+    }
+
+    const encodedValues = Array.from(value).map((v) => this.coder.encode(v));
+    return concat(encodedValues);
+  }
+
+  encode(value: InputValueOf<TCoder>): any {
```
Comment (luizstacio) [inline packages/abi-coder/src/coders/vec.ts:38]: I think the return here could be `Uint8Array`?

File: packages/contract/src/__test__/coverage-contract/coverage-contract.test.ts [hunk 5]
```diff
@@ -222,4 +227,120 @@ describe('Coverage Contract', () => {
     const { value: Some } = await contractInstance.functions.echo_option_three_u8(INPUT).call();
     expect(Some).toStrictEqual(1);
   });
+
+  it('should test u8 empty vector input', async () => {
```
Comment (luizstacio) [inline packages/contract/src/__test__/coverage-contract/coverage-contract.test.ts:231]: 💚

File: packages/providers/src/LogReader.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  toArray(): string[] {
+    return this.logs.map((log) => {
+      if (log.type === ReceiptType.LogData) {
+        const stringCoder = new StringCoder(Number(log.len));
+        let value = stringCoder.decode(arrayify(log.data), 0)[0];
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        value = value.replaceAll('\x00', '');
```
Comment (luizstacio) [inline packages/providers/src/LogReader.ts:28]: Why do we need ts-ignore here?

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/abi-coder/src/constants.ts [hunk 7]
```diff
@@ -1,7 +1,35 @@
...
+
+// TRANSACTION_SCRIPT_FIXED_SIZE = 112
+export const TRANSACTION_SCRIPT_FIXED_SIZE =
+  WORD_SIZE + // Identifier
+  WORD_SIZE + // Gas price
+  WORD_SIZE + // Gas limit
+  WORD_SIZE + // Byte price
+  WORD_SIZE + // Maturity
+  WORD_SIZE + // Script size
+  // WORD_SIZE + // Script data size
```
Comment (QuinnLee) [inline packages/abi-coder/src/constants.ts:31]: is this supposed to be commented?

**Review by QuinnLee — APPROVED** [review]

File: tsconfig.base.json [hunk 8]
```diff
@@ -4,7 +4,7 @@
     "declaration": true,
     "declarationMap": true,
     "esModuleInterop": true,
-    "lib": ["ES2020"],
+    "lib": ["ES2021"],
```
Comment (luizstacio) [inline tsconfig.base.json:7]: Do we need to change this? Any impact on the build also?

File: packages/abi-coder/src/constants.ts (hunk 7 above)
Comment (luizstacio) [reply packages/abi-coder/src/constants.ts:31]: The correct field to remove is `WORD_SIZE + // Byte price` which was removed on fuel-core v0.10.

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Review by luizstacio — APPROVED** [review]

**Review by QuinnLee — APPROVED** [review]

**Pull request merged**